dbms/src/DataStreams/MultiplexInputStream.h

```diff
@@ -0,0 +1,259 @@
+{
+private:
+    static constexpr auto NAME = "Multiplex";
+
+public:
+    MultiplexInputStream(
+        std::shared_ptr<MultiPartitionStreamPool> & shared_pool,
+        const String & req_id)
+        : log(Logger::get(NAME, req_id))
+        , shared_pool(shared_pool)
+    {
+        shared_pool->exportAddedStreams(children);
+        size_t num_children = children.size();
+        if (num_children > 1)
+        {
+            Block header = children.at(0)->getHeader();
+            for (size_t i = 1; i < num_children; ++i)
+                assertBlocksHaveEqualStructure(
+                    children[i]->getHeader(),
+                    header,
+                    "MULTIPLEX");
+        }
+    }
+
+    String getName() const override { return NAME; }
+
+    ~MultiplexInputStream() override
+    {
+        try
+        {
+            if (!all_read)
+                cancel(false);
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log, __PRETTY_FUNCTION__);
+        }
+    }
+
+    /** Different from the default implementation by trying to stop all sources,
+      * skipping failed by execution.
+      */
+    void cancel(bool kill) override
+    {
+        if (kill)
+            is_killed = true;
+
+        bool old_val = false;
+        if (!is_cancelled.compare_exchange_strong(
+                old_val,
+                true,
+                std::memory_order_seq_cst,
+                std::memory_order_relaxed))
+            return;
+
+        if (cur_stream)
+        {
+            if (IProfilingBlockInputStream * child = dynamic_cast<IProfilingBlockInputStream *>(&*cur_stream))
+            {
+                child->cancel(kill);
+            }
+        }
+    }
+
+    BlockExtraInfo getBlockExtraInfo() const override
+    {
+        throw Exception(
+            "Method getBlockExtraInfo is not supported for mode StreamUnionMode::Basic",
+            ErrorCodes::NOT_IMPLEMENTED);
+    }
+
+    Block getHeader() const override { return children.at(0)->getHeader(); }
+
+
+protected:
+    /// Do nothing, to make the preparation for the query execution in parallel, in ParallelInputsProcessor.
+    void readPrefix() override
+    {
+    }
+
+    /** The following options are possible:
+      * 1. `readImpl` function is called until it returns an empty block.
+      *  Then `readSuffix` function is called and then destructor.
+      * 2. `readImpl` function is called. At some point, `cancel` function is called perhaps from another thread.
+      *  Then `readSuffix` function is called and then destructor.
+      * 3. At any time, the object can be destroyed (destructor called).
+      */
+
+    Block readImpl() override
+    {
+        if (all_read)
+            return {};
+        if (!cur_stream)
+        {
+            cur_stream = shared_pool->pickOne();
+            if (!cur_stream)
+            { // shared_pool is empty
+                all_read = true;
+                return {};
+            }
```

Comment:
do we need `cur_stream->readPrefix()` here?

Reply:
readPrefix( is called in L232

Reply:
if `MultiplexInputStream.cur_stream == nullptr` and `!shared_pool.empty()`, `readImpl` would be
- L206 `cur_stream = shared_pool->pickOne()`
- L215 `while (!(ret = cur_stream->read()))`
`cur_stream->readPrefix()` is not called before `cur_stream.read()`

Reply:
good spot

Reply:
updated